### PR TITLE
fix: catalog name confusion and overwrite of same assembled files

### LINKED ISSUE #1073: Converting catalog to markdown and back to json looses catalog title
## Describe the bug

Taking a json catalog (such as the [800-53r4 reference catalog](https://github.com/usnistgov/oscal-content/blob/main/nist.gov/SP800-53/rev4/json/NIST_SP-800-53_rev4_catalog.json)) and converting to markdown, and then back to json causes the catalog title to be replaced with 'REPLACE_ME':

## To Reproduce

1. mkdir trestletest
1. cd trestletest/
1. . ../../compliance-trestle/venv.trestle-develop/bin/activate
1. trestle init
1.  trestle import -f ../oscal-content/nist.gov/SP800-53/rev4/json/NIST_SP-800-53_rev4_catalog.json -o NIST_SP-800-53_rev4
1. trestle author catalog-generate --name NIST_SP-800-53_rev4 --output md_catalogs/NIST_800-53_rev4
1. trestle author catalog-assemble -v --set-parameters --markdown md_catalogs/NIST_800-53_rev4 --output NIST_SP-800-53_rev4
1. head catalogs/NIST_SP-800-53_rev4/catalog.json 

```
head catalogs/NIST_SP-800-53_rev4/catalog.json 
{
  "catalog": {
    "uuid": "fd2eac2a-1ec9-4f66-990e-9dcecab12055",
    "metadata": {
      "title": "REPLACE_ME",
      "last-modified": "2022-03-29T15:22:35.697670+01:00",
      "version": "REPLACE_ME",
      "oscal-version": "1.0.0"
    },
    "controls": [],
```

## Expected behavior

```
head ../oscal-content/nist.gov/SP800-53/rev4/json/NIST_SP-800-53_rev4_catalog.json 
{
  "catalog": {
    "uuid": "b954d3b7-d2c7-453b-8eb2-459e8d3b8462",
    "metadata": {
      "title": "NIST Special Publication 800-53 Revision 4: Security and Privacy Controls for Federal Information Systems and Organizations",
      "last-modified": "2021-06-08T13:57:28.91745-04:00",
      "version": "2015-01-22",
      "oscal-version": "1.0.0",
      "props": [
        {
```

### PR DESCRIPTION
**Pull request by fsuits**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [x] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Summary
Simpler name handling in catalog assemble
All assemble tools should overwrite only when needed
Removed assumption of json output in assemble tools
## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

closes #1073

### CODE REVIEW

File: trestle/core/commands/author/profile.py [hunk 1]
```diff
@@ -312,16 +311,29 @@ def assemble_profile(
...
+            parent_prof_name = assem_prof_name
+
+        try:
+            parent_prof, parent_prof_path = ModelUtils.load_top_level_model(
+                trestle_root,
+                parent_prof_name,
+                prof.Profile
+            )
+        except Exception as e:
+            raise TrestleError(f'Error loading original profile {parent_prof_name}: {e}')
```
Comment (enikonovad) [inline trestle/core/commands/author/profile.py:335]: Let the original exception fall up the stack

File: trestle/core/commands/author/catalog.py [hunk 2]
```diff
@@ -174,55 +175,75 @@ def assemble_catalog(
...
+        new_content_type = FileContentType.JSON
+
+        # if we have parent catalog then merge the markdown controls into it
+        # the parent can be a separate catalog or the destination assembled catalog if it exists
+        # but this is the catalog that the markdown is merged into in memory
+        if parent_cat_name:
             try:
-                orig_cat, _ = ModelUtils.load_top_level_model(trestle_root, orig_cat_name, Catalog)
+                parent_cat, parent_cat_path = ModelUtils.load_top_level_model(trestle_root, parent_cat_name, Catalog)
             except Exception as e:
```
Comment (enikonovad) [inline trestle/core/commands/author/catalog.py:210]: Let the original exception fall up the stack, or at least catch specific exception

File: trestle/core/commands/author/catalog.py [hunk 3]
```diff
@@ -174,55 +175,75 @@ def assemble_catalog(
...
+        assem_cat_path = ModelUtils.path_for_top_level_model(trestle_root, assem_cat_name, Catalog, new_content_type)
+
+        if assem_cat_path.parent.exists():
             logger.info('Creating catalog from markdown and destination catalog directory exists, so overwriting.')
             try:
-                shutil.rmtree(str(new_cat_dir))
+                shutil.rmtree(str(assem_cat_path.parent))
             except OSError as e:
-                raise TrestleError(f'OSError deleting existing catalog directory with rmtree {new_cat_dir}: {e}')
+                raise TrestleError(
```
Comment (enikonovad) [inline trestle/core/commands/author/catalog.py:241]: This doesn't add any value to the user or handling the error, please remove it and let it fall

File: trestle/core/commands/author/catalog.py [hunk 4]
```diff
@@ -174,55 +175,75 @@ def assemble_catalog(
...
+                    f'OSError deleting existing catalog directory with rmtree {assem_cat_path.parent}: {e}'
+                )
         try:
-            new_cat_dir.mkdir()
-            md_catalog.oscal_write(new_cat_dir / 'catalog.json')
+            assem_cat_path.parent.mkdir()
+            md_catalog.oscal_write(assem_cat_path.parent / 'catalog.json')
         except OSError as e:
-            raise TrestleError(f'OSError writing catalog from markdown to {new_cat_dir}: {e}')
```
Comment (enikonovad) [inline trestle/core/commands/author/catalog.py:227]: Ditto

File: trestle/core/commands/author/profile.py [hunk 5]
```diff
@@ -336,47 +348,38 @@ def assemble_profile(
...
                 return CmdReturnCodes.SUCCESS.value
 
         if regenerate:
-            new_profile, _, _ = ModelUtils.regenerate_uuids(new_profile)
-        ModelUtils.update_last_modified(new_profile)
+            parent_prof, _, _ = ModelUtils.regenerate_uuids(parent_prof)
+        ModelUtils.update_last_modified(parent_prof)
 
-        if new_profile_dir.exists():
+        if assem_prof_path.parent.exists():
```
Comment (enikonovad) [inline trestle/core/commands/author/profile.py:372]: Will fail if assem_prof_path doesn't exist

File: trestle/core/commands/author/profile.py [hunk 6]
```diff
@@ -336,47 +348,38 @@ def assemble_profile(
...
-                no_changes = False
-            # at this point if no known changes then do the full and possibly expensive comparison to confirm
-            if no_changes and existing_profile == new_profile:
+            parent_prof.metadata.version = com.Version(__root__=version)
+
+        assem_prof_path = ModelUtils.path_for_top_level_model(
+            trestle_root, assem_prof_name, prof.Profile, new_content_type
+        )
+
+        if assem_prof_path.exists():
```
Comment (enikonovad) [inline trestle/core/commands/author/profile.py:362]: Better to check for non existence and raise exception if the case

File: trestle/core/commands/author/profile.py [hunk 7]
```diff
@@ -336,47 +348,38 @@ def assemble_profile(
...
 
-        if new_profile_dir.exists():
+        if assem_prof_path.parent.exists():
             logger.info('Creating profile from markdown and destination profile directory exists, so updating.')
             try:
-                shutil.rmtree(str(new_profile_dir))
+                shutil.rmtree(str(assem_prof_path.parent))
             except OSError as e:
-                raise TrestleError(f'OSError deleting existing catalog directory with rmtree {new_profile_dir}: {e}')
+                raise TrestleError(
```
Comment (enikonovad) [inline trestle/core/commands/author/profile.py:377]: No value in this for the user

File: trestle/core/commands/author/profile.py [hunk 8]
```diff
@@ -336,47 +348,38 @@ def assemble_profile(
...
+                    f'OSError deleting existing catalog directory with rmtree {assem_prof_path.parent}: {e}'
+                )
         try:
-            new_profile_dir.mkdir()
-            new_profile.oscal_write(new_profile_dir / 'profile.json')
+            assem_prof_path.parent.mkdir()
+            parent_prof.oscal_write(assem_prof_path)
         except OSError as e:
-            raise TrestleError(f'OSError writing profile from markdown to {new_profile_dir}: {e}')
+            raise TrestleError(f'OSError writing profile from markdown to {assem_prof_path.parent}: {e}')
```
Comment (enikonovad) [inline trestle/core/commands/author/profile.py:384]: ditto

**Review by enikonovad — CHANGES_REQUESTED** [review]
Overall looks good, some comments. I think since we are touching this code, it would be nice to refactor the exception handling.

File: trestle/core/commands/author/profile.py (hunk 5 above)
Comment (fsuits) [reply trestle/core/commands/author/profile.py:372]: I think it's ok - existence isn't required to know parent.  It's done by operating on the path as a string.
Example:
In empty but existing directory `/tmp/farble`:
```
>>> import pathlib
>>> p = pathlib.Path('/tmp/farble/foo/bar/cow.txt')
>>> p
WindowsPath('/tmp/farble/foo/bar/cow.txt')
>>> p.exists()
False
>>> p.parent.exists()
False
>>> p.parent
WindowsPath('/tmp/farble/foo/bar')
>>> p.parent.mkdir(parents=True)
>>> p
WindowsPath('/tmp/farble/for/bar/cow.txt')
```

File: trestle/core/commands/author/profile.py (hunk 6 above)
Comment (fsuits) [reply trestle/core/commands/author/profile.py:362]: This is ok - it's just loading the file if it exists.  It doesn't need to exist.

**Comment by fsuits** [comment]
I addressed all exception requests (happy to remove code) but did not make changes for the two items I commented.  Please confirm those two items are ok.

**Review by enikonovad — APPROVED** [review]
LGTM, thanks for the changes

**Pull request merged**